### ISSUE TITLE
iterable_contains_unrelated_type not working for List

### ISSUE THREAD

**Issue by davidmorgan** [dart-lang/linter#1003]
In IntelliJ, I see this triggers the lint:

```
  Iterable<int> foo = [1];
  print(foo.contains('3'));
```

and this:

```
  Set<int> foo = Set.from([1]);
  print(foo.contains('3'));
```

but not this:

```
  List<int> foo = [1];
  print(foo.contains('3'));
```

I'm at a loss as to why not; it looks like there are test cases that should catch this in the linter code, and they pass. Maybe there is something in the test setup that is now different from the actual `List`?

**Comment by alexeieleusis** [comment]
I just tried adding the example you provide and the test fails, also confirmed that if I remove the `// LINT` annotation the test fails, so is definitely being linted in the tests. If I run `dart bin/linter.dart --rules=iterable_contains_unrelated_type test/rules/iterable_contains_unrelated_type.dart` the lint is not reported.

I have no idea what the problem is :(

**Comment by pq** [comment]
> Maybe there is something in the test setup that is now different from the actual List?

👍 

We mock the SDK to speed up testing and it looks like `List` is out of date:


https://github.com/dart-lang/linter/blob/9f3df652baa88372513f1ac74148b8e7a5a6d1a0/test/mock_sdk.dart#L125-L134


(In the SDK, `List` implements `EfficientLengthIterable<E>`)

**Comment by lambda-fairy** [comment]
I think the issue is that [`unrelated_types_visitor` only checks `interfaces`](https://github.com/dart-lang/linter/blob/6998e5b65cf51250315c2a3293d2a02af2858ff6/lib/src/util/unrelated_types_visitor.dart#L28), and not `superclass` or `mixins`.

Indeed, in the SDK [`List<T>` *extends* `EfficientLengthIterable`](https://github.com/dart-lang/sdk/blob/86edfbef87b5207a48a0eff386f4dd48bf320f18/sdk/lib/internal/iterable.dart#L30), not implements.

Perhaps replacing that bespoke method with a call to [`getImplementedInterfaces`](https://github.com/dart-lang/linter/blob/6998e5b65cf51250315c2a3293d2a02af2858ff6/lib/src/util/dart_type_utilities.dart#L161-L180) would fix the bug.

**Comment by lambda-fairy** [comment]
I take that back -- after changing the mock `List<T>` to extend an intermediate `EfficientLengthIterable` class, the tests still pass 🤔

NB: The mock SDK has been moved to the [analyzer package](https://github.com/dart-lang/sdk/blob/c10b41a4e2da04406ff07d8cb7ee14e38ff13d1c/pkg/analyzer/lib/src/test_utilities/mock_sdk.dart).

**Comment by xvrh** [comment]
I think the lint `iterable_contains_unrelated_type` is still broken. The examples don't trigger the lint in my project.

```
dart --version
Dart SDK version: 2.12.1 (stable) (Wed Mar 10 10:18:47 2021 +0100) on "macos_x64"
```

```dart
// main.dart
void someFunction() {
  var list = <int>[];
  if (list.contains('1')) print('someFunction');
}
```

```yaml
# analysis_options.yaml
linter:
  rules:
    - iterable_contains_unrelated_type
```

```
dart analyze
No issues found!
```
